src/rng.rs

```diff
@@ -0,0 +1,374 @@
+        } else {
+            Ok(output)
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[test]
+        fn parse_seed() {
+            assert_eq!("[0, 0, 0, 0]".parse(), Ok(Seed([0, 0, 0, 0])));
+            assert_eq!("[0, 1, 2, 3]".parse(), Ok(Seed([0, 1, 2, 3])));
+            assert_eq!(
+                "[2173726344, 4077344496, 2175816672, 3385125285]".parse(),
+                Ok(Seed([
+                    2_173_726_344,
+                    4_077_344_496,
+                    2_175_816_672,
+                    3_385_125_285
+                ]))
+            );
+            assert_eq!("".parse(), Err::<Seed, _>(ParseError));
+        }
+    }
+}
+
+#[cfg(feature = "mock_base")]
+mod seed_printer {
+    use super::Seed;
+    use std::thread;
+
+    /// Helper struct that prints a seed on scope exit.
+    pub struct SeedPrinter {
+        seed: Seed,
+        mode: Mode,
+    }
+
+    #[derive(Eq, PartialEq)]
+    enum Mode {
+        OnFailure,
+        OnSuccess,
+    }
+
+    impl SeedPrinter {
+        /// Create new `SeedPrinter` that will print the given seed on failure.
+        pub fn on_failure(seed: Seed) -> Self {
+            interrupt::activate(seed);
+
+            Self {
+                seed,
+                mode: Mode::OnFailure,
+            }
+        }
+
+        /// Create new `SeedPrinter` that will print the given seed on success (exiting the scope
+        /// without panic).
+        pub fn on_success(seed: Seed) -> Self {
+            Self {
+                seed,
+                mode: Mode::OnSuccess,
+            }
+        }
+
+        /// Returns the seed this printer will print.
+        pub fn seed(&self) -> &Seed {
+            &self.seed
+        }
+    }
+
+    impl Drop for SeedPrinter {
+        fn drop(&mut self) {
+            interrupt::deactivate();
+
+            if thread::panicking() {
+                if self.mode == Mode::OnFailure {
+                    print_seed(&self.seed, "");
+                }
+            } else if self.mode == Mode::OnSuccess {
+                print_seed(&self.seed, "");
+            }
+        }
+    }
+
+    fn print_seed(seed: &Seed, label: &str) {
+        let msg = if label.is_empty() {
+            format!("{}", seed)
+        } else {
+            format!("{}: {}", label, seed)
+        };
+
+        let border = (0..msg.len()).map(|_| "=").collect::<String>();
+        eprintln!("\n{}\n{}\n{}\n", border, msg, border);
+    }
+
+    // Print the seed also on SIGINT and SIGTERM
+    mod interrupt {
+        use super::{print_seed, Seed};
+        use lazy_static::lazy_static;
+        use std::{
+            collections::HashMap,
+            process,
+            sync::{Mutex, Once},
+            thread::{self, ThreadId},
+        };
+
+        static HANDLER_SET: Once = Once::new();
+
+        lazy_static! {
+            static ref ACTIVE_SEEDS: Mutex<HashMap<ThreadId, (String, Seed)>> =
```

Comment:
Why is there a HashMap? I thought before we only had a single seed running, is there a reason for that?

Reply:
There is no single global seed anymore and there can be multiple tests running in parallel. Also, this functionality is only to support interruptions (Ctrl+C, etc...).